front/types/assistant/conversation.ts

```diff
@@ -0,0 +1,119 @@
+import { ModelId } from "@app/lib/models";
+
+import { UserProviderType, UserType } from "../user";
+
+/**
+ * Mentions
+ */
+
+export type AssistantAgentMention = {
+  assistantId: string;
+};
+
+export type AssistantUserMention = {
+  provider: UserProviderType;
+  providerId: string;
+};
+
+export type AssistantMention = AssistantAgentMention | AssistantUserMention;
+
+/**
+ * User messages
+ */
+
+export type AssistantUserMessageType = {
+  id: ModelId;
+  sId: string;
+  status: "visible" | "deleted";
+  parentMessageId: string;
+  user?: UserType;
+  mentions: AssistantMention[];
+  message: string;
+  context: {
+    username: string;
+    timezone: string;
+    fullName?: string;
+    email?: string;
+    profilePictureUrl?: string;
+  };
+};
+
+/**
+ * Retrieval action
+ */
+
+export type RetrievalDocumentType = {
+  id: ModelId;
+  dataSourceId: string;
+  sourceUrl?: string;
+  documentId: string;
+  timestamp: number;
+  tags: string[];
+  score: number;
+  chunks: {
+    text: string;
+    offset: number;
+    score: number;
+  }[];
+};
+
+export type RetrievalActionType = {
+  id: ModelId;
+  params: {
+    query: string;
+  };
+  documents: RetrievalDocumentType[];
+};
+
+/**
+ * Agent messages
+ */
+
+export type AssistantUserFeedbackType = {
+  user: UserType;
+  value: "positive" | "negative" | null;
+  comment?: string;
+};
+
+export type AssistantAgentActionType = RetrievalActionType;
+
+/**
+ * Both `action` and `message` are optional (we could have a no-op agent basically).
+ *
+ * Since `action` and `message` are bundled together, it means that we will only be able to retry
+ * them together in case of error of either. We store an error only here whether it's an error
+ * coming from the action or from the message generation.
+ */
+export type AssistantAgentMessageType = {
+  id: ModelId;
+  sId: string;
+  status: "visible" | "deleted";
+  parentMessageId: string;
+  action?: AssistantAgentActionType;
+  message?: string;
+  feedbacks: AssistantUserFeedbackType[];
+  error?: {
+    code: string;
+    message: string;
+  };
+};
+
+/**
+ * Conversations
+ */
+
+export type AssistantConversationVisibility = "private" | "workspace";
+
+/**
+ * content [][] structure is intended to allow retries (of agent messages) or edits (of user
+ * messages).
+ */
+export type AssistantConversationType = {
+  id: ModelId;
+  created: number;
+  sId: string;
+  title?: string;
+  participants: UserType[];
+  content: (AssistantUserMessageType[] | AssistantAgentMessageType[])[];
```

Comment:
We are going to need some magic flag to know what type it is here at runtime.
Otherwise we would need a class if we want to do it without any magic flag.
Just mentioning it :)

Reply:
Yeah we'll likely iterate on that if this is too much of an issue. I presume here that we can use a isAssistantUserMessage and isAssistantAgentMessage functions similar to Result, but maybe I'm wrong?

Reply:
We can use functions like this to do type assertion, but we'll need to rely on some known properties of the flag and set it for each object of these types we instantiate. Exactly the same way we do it with `__is_dust_error` here:
```
 const workflowError: WorkflowError = {
      type: "transient_upstream_activity_error",
      message:
        "Received unexpected undefined replies from Slack API in getMessagesForChannel (generally transient)",
      __is_dust_error: true,
    };
```

Reply:
Yep. I thiiiink this should be easy-ish we'll have clear invariants that separate the two (we can add a string to the type to name it and use that as a flag also)?

Sounds good to you?